Add multicore implementation of `ttnn.experimental.cumsum()`

nmauriceTT opened the pull request:
### Ticket
#23117 

### Problem description


A single core implementation of ttnn.experimental.cumsum() was added in https://github.com/tenstorrent/tt-metal/pull/21196

While this implementation supports more ranks (<4 and > 4) and dim parameters than moreh_cumsum(), it is currently single-threaded.

### What's changed

This PR adds:

- A multi-core implementation of `ttnn.experimental.cumsum()` (single device)
- Support for `override_runtime_arguments()`
- New tests for tensor rank up to 8, and on each axis (c.f. #14549 and #22865 ).
- Check against memory overhead (c.f. #23264 ): logs a warning if more intermediate tensor is more than twice as large as input, and throws if >800MiB overhead. Proper resolution of memory issue is planned in an upcoming PR. 

Kernels can also take a `flip` argument, but full support will be added later (for eventual merger with moreh's cumsum) to keep the scope of this PR more reasonable. 

#### Implementation details 

Parallelization is done on independent rows. No inter-core communication or synchronization is required. 

#### Performance analysis 

(With all 64 cores on wormhole)

With (hyper)-cubic tensors (N in `{16, 32, 38, 64, 96, 128, 160}`)

![benchmark-cumsum-parallel-tracy3_BFLOAT16](https://github.com/user-attachments/assets/623eb94a-9de3-43b8-94d4-b96c7e4269d0)

Note: `dim=2` and `dim=3` are slower due to their pre and post-processing (`ttnn::permute`). This is something we would like to remove.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [X] New/Existing tests provide coverage for changes

nmauriceTT committed:
Linearize i0/i1 in dataflow kernels

nmauriceTT committed:
cumsum: 'multicore' work with ncores=1

nmauriceTT committed:
cumsum: basic parallel cumsum

nmauriceTT committed:
cumsum: fix mismatched circular buffer index

nmauriceTT committed:
cumsum: add cumsum_bw() (untested)

nmauriceTT committed:
cumsum: add override_runtime_arguments and hash computation (WIP)

nmauriceTT committed:
cumsum: increase size of circular buffer to more than 1 tile

nmauriceTT committed:
cumsum: add long tensor tests

nmauriceTT committed:
cumsum: handle override_runtime_arguments

nmauriceTT committed:
cumsum: add test for cumsum_backward (non-passing)

nmauriceTT committed:
cumsum: add test for cumsum_backward and forward moreh_cumsum to ttnn.experimental.cumsum

nmauriceTT committed:
cumsum: fixed backward cumsum

nmauriceTT committed:
Fix runtime error whenn calling override_runtime_arguments with less thant 64 threads

nmauriceTT committed:
Update copyright header and cleanup code

nmauriceTT committed:
cumsum: add tests for 6D tensors

nmauriceTT committed:
cumsum: remove ttnn.experimental.cumsum_backwared (will be add in another PR)

nmauriceTT committed:
cumsum: update copyright header

nmauriceTT committed:
cumsum: Remove print statements

nmauriceTT committed:
cumsum: Add tests with long tensors

nmauriceTT committed:
Revert provisional edits on moreh files

nmauriceTT committed:
Cleanup references to flip and mmeory_config parameters: to be re-added later

nmauriceTT committed:
cumsum: log warning if 2x tensor overhead and TT_FATAL() if overhead > 800 MiB

nmauriceTT committed:
cumsum: move some kernel parameters to compile time args

nmauriceTT committed:
cumsum: replace assert_with_pcc with assert_allclose

nmauriceTT committed:
cumsum: constants TILE_HW => tile hw from tensor specs

nmauriceTT committed:
cumsum: Add tests with program cache

nmauriceTT committed:
cumsum: print warning before permute based on both relative and absolute size increase

nmauriceTT committed:
cumsum: cleanup

nmauriceTT committed:
cumsum: fix constexpr-related compilation error

nmauriceTT committed:
cumsum: Fix test with program cache

bbradelTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumsum/device/kernels/dataflow/cumsum_reader.cpp:
```diff
@@ -46,21 +50,27 @@ void kernel_main() {
-            for (unsigned j = 0; j < tiles_per_row; j++) {
-                uint32_t tileid = get_tile_id(i0, i1, j, tiles_per_row, product_low_dims, product_high_dims, HtWt);
+    for (uint32_t i = start_row; i < start_row + num_rows; i++) {
+        uint32_t i0 = i / (product_high_dims * HtWt);
```
Could you please change the code to avoid / and % in the kernels?

nmauriceTT committed:
Fix according to PR feedback

- add proper pytest skips
- remove integer division from kernels
- fix 'underflow' problem with tensor_rank
- use `tensor_shape[-1]` instead of `tensor_shape[tensor_rank - 1]`
- remove confusing comment

nmauriceTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumsum/device/cumsum_program_factory.cpp:
```diff
@@ -82,48 +83,102 @@ CumSumDeviceOperation::SingleCore::cached_program_t CumSumDeviceOperation::Singl
+    const auto& tile = input_tensor.tensor_spec().tile();
+    uint32_t num_tiles = output_tensor.physical_volume() / tile.get_tile_hw();
+
+    const uint32_t xy_volume = tensor_shape[tensor_rank - 1] * tensor_shape[tensor_rank - 2];  // W * H
```
Yes. I changed it.

nmauriceTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumsum/cumsum.cpp:
```diff
@@ -88,6 +115,52 @@ Tensor CumSumOperation::invoke(
+            constexpr uint64_t EIGHT_KB = (1024 * 4);
+            constexpr uint64_t MAX_ALLOWED_GROWTH = 800 * ONE_MB;  // 800 MiB
+
+            TT_ASSERT(old_volume > 0, "Can not compute permuted tensor for cumsum operation if input is empty");
```
I changed it to `TT_FATAL()`

nmauriceTT commented on tests/ttnn/unit_tests/operations/reduce/test_cumsum.py:
```diff
@@ -166,6 +187,9 @@ def test_cumsum_with_preallocated_output(size, dim, dtypes, device):
     if expected_output_dtype == ttnn.int32 or expected_output_dtype == ttnn.uint32:
         return
 
+    if not is_supported(size, dim, expected_output_dtype):
```
Done. I've added a `pytest.skip("Unsupported configuration by ttnn.experimental.cumsum")` skip instead of just a return.

nmauriceTT committed:
cumsum: remove leftover division + add short comment
